File: python/grid_filter/kdtree_util.py
""" Utility and helper functions for KD-Tree building
"""
from typing import Tuple, List

def sort_points(pts: List[Tuple[float, float, int]], dim: int) -> List[Tuple[float, float, int]]:
    """Return a list of sorted points by dimension.

    Keyword arguments:
    pts -- list of (lat, lon, cell_index) points
    dim -- dimension to sort over. (0 : lat, 1 : lon)
    """
    if dim == 0:
        pts = sorted(pts, key=lambda pt: pt[0])
    elif dim == 1:
        pts = sorted(pts, key=lambda pt: pt[1])
    else:
        raise RuntimeError("Could not sort on index: {dim}".format(dim=dim))
    return pts

def median_point_id(pts: List[Tuple[float, float, int]]) -> Tuple[Tuple[float, float, int], int]:
    """Return the index, point of the median point in list of sorted points.

    Keyword arguments:
    pts -- list of (lat, lon, cell_index) points
    dim -- dimension to search for media (0 : x, 1 : y)
    """
    idx = int((len(pts) - 1) // 2)

    return pts[idx], idx

def euclidean_2d_distance_sq(p1: Tuple[float, float], p2: Tuple[float, float]) -> float:
    """Calculate the Euclidean squared distance between 2D points, p1 and p2"""
    return (p2[0]-p1[0])**2 + (p2[1]-p1[1])**2

def euclidean_1d_distance_sq(p1: Tuple[float, float], p2: Tuple[float, float], dim) -> float:
    """Calculate the x- or y- projected squared distance between two points"""
    if (dim == 0) or (dim == 1):
        dist_sq = (p2[dim] - p1[dim])**2
    else:
        raise RuntimeError(f'Invalid dimension: {dim}')
    return dist_sq

File: src/kdtree_util.cpp
/*
 * File - kdtree_util.cpp
 * KD Tree utility functions.
 */
#include <iostream>
#include <memory>
#include <cmath>
#include "kdtree_node.h"
#include "kdtree_util.h"

// Compare latitude component of nodeData
bool compare_lat(nodeData d1, nodeData d2) {
  return (d1.lat < d2.lat);
}

// Compare longitude component of nodeData
bool compare_lon(nodeData d1, nodeData d2) {
  return (d1.lon < d2.lon);
}

// Return the median data and index of a list
std::tuple<nodeData, int> median_point_id(const std::vector<nodeData> &nd) {
  int median_id = floor(nd.size()/2);
  std::tuple<nodeData, int> median_node{nd.at(median_id), median_id};
  return std::tuple<nodeData, int>(median_node);
}

// Projected distance (squared) between 2D points in given dimension.
double euclidean_1d_distance_sq(point2D p1, point2D p2, int dim) {
  double dist_sq = 0.0;
  if (dim == 0) {
    dist_sq = pow((p2.lat - p1.lat),2.0);  
  } else if (dim == 1) {
    dist_sq = pow((p2.lon - p1.lon),2.0);
  } else {
    dist_sq = -1.0; 
  }
  return dist_sq;
}

// Distance (squared) between 2D points.
double euclidean_2d_distance_sq(point2D p1, point2D p2) {
  return pow((p2.lat-p1.lat),2.0) + pow((p2.lon-p1.lon),2.0);
}
